--- gnuboy-1.0.3/sys/x11/xlib.c	Sun Sep 23 21:01:10 2001
+++ gnuboy-1.0.3-nk/sys/x11/xlib.c	Tue Oct 26 12:09:27 2004
@@ -60,24 +60,8 @@ static char *x_displayname;
 static Display *x_display;
 static int x_screen;
 
-static struct
-{
-	int bits;
-	int vc;
-	int bytes;
-} x_vissup[] =
-{
-	{ 8, PseudoColor, 1 },
-	{ 15, TrueColor, 2 },
-	{ 16, TrueColor, 2 },
-	{ 32, TrueColor, 4 },
-	{ 24, TrueColor, 3 },
-	{ 0, 0, 0 }
-};
-
 static int x_bits, x_bytes;
 static Visual *x_vis;
-static XVisualInfo x_visinfo;
 static int x_pseudo;
 static Colormap x_cmap;
 static XColor x_ctable[256];
@@ -94,7 +78,7 @@ static GC x_gc;
 
 static XSizeHints x_size;
 static XWMHints x_wmhints;
-/*static XClassHint x_class;*/
+static XClassHint x_class;
 
 #ifdef USE_XSHM
 static XShmSegmentInfo x_shm;
@@ -247,25 +231,19 @@ void vid_init()
 		die("failed to connect to X display\n");
 	x_screen = DefaultScreen(x_display);
 
-	for (i = 0; x_vissup[i].bits; i++)
-	{
-		if (XMatchVisualInfo(
-			x_display, x_screen,
-			x_vissup[i].bits, x_vissup[i].vc, &x_visinfo))
-		{
-			if (x_vissup[i].vc == PseudoColor)
-				x_pseudo = 1;
-			else
-				x_pseudo = 0;
-			x_bits = x_vissup[i].bits;
-			x_bytes = x_vissup[i].bytes;
-			break;
-		}
-	}
-	if (!x_bits) die("no suitable X visuals\n");
-	x_vis = x_visinfo.visual;
+	x_vis = DefaultVisual(x_display, x_screen);
 	if (!x_vis) die("X visual is NULL");
 
+	x_cmap = DefaultColormap(x_display, x_screen);
+	x_bits = DefaultDepth(x_display, x_screen);
+	x_bytes = x_bits >> 3;
+
+	/* NOTE: This code assumes that all 8-bit color displays are PseudoColor.
+	 * This assumption is correct almost everywhere. */
+
+	if (x_bytes == 1)
+		x_pseudo = 1;
+	
 	if (x_pseudo)
 	{
 		x_cmap = XCreateColormap(
@@ -324,8 +302,10 @@ void vid_init()
 	x_wmhints.flags = StateHint | InputHint;
 	XSetWMHints(x_display, x_win, &x_wmhints);
 
-	/* FIXME - set X class info stuff (with XSetClassHint)... */
-	
+	x_class.res_name = "xgnuboy";
+	x_class.res_class = "xgnuboy";
+	XSetClassHint(x_display, x_win, &x_class);
+
 	XMapWindow(x_display, x_win);
 
 	for(;;)

